fix(font): Apply glyph constraints before thickening and centering before quantizing

danielwe opened the pull request:
In CoreText, when thickening (font smoothing) is enabled or Ghostty is synthesizing a bold face, the glyph bounding box is padded to make sure the thicker glyph can fit. Currently, this happens before applying constraints (scaling and alignment), which makes the size and position of constrained glyphs dependent on font size, font thickening strength, and display DPI.

With this PR, constraints are applied before any other adjustments, and padding is applied directly to the rasterization canvas without modifying any metrics.

For consistency, I also moved constraint application above emboldening in the FreeType code, although under that API, the two operations are orthogonal as far as I can tell.

Secondly, this PR moves glyph centering above bitmap quantization, as centering is generally fractional and will therefore undo the quantizing if done after.

Supersedes #8552.

danielwe committed:
Center before quantizing bitmap glyphs

mitchellh commented:
Merged some of your other PRs first which caused a conflict, can you please rebase this one.

danielwe commented:
Rebased and ready. Made a tiny change to avoid an unnecessary merge conflict between this #8847.

qwerasd205 commented on src/font/face/coretext.zig:
```diff
@@ -400,22 +383,32 @@ pub const Face = struct {
         // our drawn glyph, including the fractional offset.
-        const px_width: u32 = @intFromFloat(@ceil(width + frac_x));
-        const px_height: u32 = @intFromFloat(@ceil(height + frac_y));
+        const px_width: u32 = @intFromFloat(@ceil(width + frac_x + canvas_padding));
+        const px_height: u32 = @intFromFloat(@ceil(height + frac_y + canvas_padding));
```
This should be adding `2 * canvas_padding` since the width and height need the padding added to both edges.

danielwe replied:
Fair! I'll include that change in the next push.

danielwe replied:
No, wait a second, the problem is that `canvas_padding` isn't necessarily an integer either, so we'd have to split that too into a rounded and fractional part. That seems more complicated. Maybe we could rename the variables to make the current flow more intuitive?

danielwe replied:
Though if we move synthetic bold pre-constraint as discussed below, the padding is always integral, so this problem goes away.

qwerasd205 reviewed (commented):
Why did you move the code that bails out for glyphs that are too small after all the (relatively) expensive constraint calculations?

Also, as a side note, I'm pretty sure if we're gonna move the FreeType outline emboldening code at all it should be moved *before* we measure the glyph (additional safety checks may need to be added to make sure that there are outlines to embolden in the first place).

qwerasd205 commented:
Font thickening and emboldening are completely different things.

qwerasd205 commented:
I see your argument about bold symbols having the same basic proportions as non-bold ones with synthetic bold, but that's not the effect we'd get if a font *did* have a bold with symbols that have constraints applied, and it feels odd to me to be inconsistent about that.

danielwe committed:
Use correct and consistent pre-constraint glyph rect

In Freetype, measure rect after emboldening, so constraints apply to the
true glyph size like in CoreText.

In CoreText, don't let font smoothing affect the rect (only the canvas).

danielwe commented:
Alright, I pushed a new version that makes a simpmler and more minimal set of changes along the lines you indicated.
* In FreeType, we measure the possibly emboldened glyph, which is consistent with the rect padding done in CoreText. Thus, both backends apply constraints to the emboldened bbox.
* We keep the main change: only padding the canvas, not the rect, for font smoothing. Since this padding is always integer (no fractional contribution from emboldening), we keep it separate from `frac_{x,y}`.

danielwe committed:
Fix comment

danielwe commented:
Sounds great!

I'll note that #8990 and #8847 address higher-severity bugs/regressions, so maybe they should take priority if the time window for 1.2.1 is getting tight. That said, this one is probably the quickest and easiest to review.
